snuba/pipeline/pipeline_delegator.py

```diff
@@ -0,0 +1,137 @@
+    and storages in order to test and compare the results and performance of
+    various pipelines before rolling them out to users.
+
+    The pipelines that are built and executed on a particular query are determined
+    by the selector function which returns a sequence of query builder IDs. The
+    first item in the list corresponds to the pipeline that will be used to ultimately
+    return a result. Hence the selector function must return a list with at least
+    one value in it.
+
+    Subsequent (if any) pipelines that are also returned by the selector function
+    will be executed in separate threads. The results of these other queries are
+    provided to the callback function if one is provided.
+
+    This component is produced by the PipelineDelegator.
+    """
+
+    def __init__(
+        self,
+        request: Request,
+        runner: QueryRunner,
+        query_pipeline_builders: QueryPipelineBuilders,
+        selector_func: SelectorFunc,
+        callback_func: Optional[CallbackFunc],
+    ):
+        self.__request = request
+        self.__runner = runner
+        self.__query_pipeline_builders = query_pipeline_builders
+        self.__selector_func = selector_func
+        self.__callback_func = callback_func
+
+    def execute(self) -> QueryResult:
+        def build_and_execute_pipeline(
+            query_pipeline_builder: QueryPipelineBuilder,
+        ) -> QueryResult:
+            return query_pipeline_builder.build_pipeline(
+                self.__request, self.__runner
+            ).execute()
+
+        def execute_pipelines() -> Iterator[Tuple[BuilderId, QueryResult]]:
+            builder_ids = self.__selector_func(self.__request.query)
+            primary_builder_id = builder_ids.pop(0)
+
+            futures = [
+                (
+                    builder_id,
+                    executor.submit(
+                        build_and_execute_pipeline,
+                        query_pipeline_builder=self.__query_pipeline_builders[
+                            builder_id
+                        ],
+                    ),
+                )
+                for builder_id in builder_ids
+            ]
+
+            yield primary_builder_id, build_and_execute_pipeline(
+                self.__query_pipeline_builders[primary_builder_id]
+            )
+
+            yield from [
+                (builder_id, future.result()) for (builder_id, future) in futures
+            ]
+
+        generator = execute_pipelines()
+
+        query_results: QueryResults = []
+
+        try:
+            builder_id, query_result = next(generator)
+            query_results.append((builder_id, query_result))
+            return query_result
+        finally:
+            try:
+                for builder_id, query_result in generator:
+                    query_results.append((builder_id, query_result))
+
+                if self.__callback_func is not None:
+                    self.__callback_func(query_results)
```

Comment:
I assume you did this in order to return immediately when the primary pipeline returns and not waiting for the second one. If this is correct, are you sure that python will not wait for the `finally` block to complete before returning, and if it returns right away, who and when executes the finally block ?

Reply:
Actually, this is a mistake, `finally` will finish before returning which is not what we want. I change this to do the finally part in a thread now to avoid this, though I'm not sure if there's a neater way to achieve this.

Reply:
Ok, but to be sure we are on the same page, what is the behavior you would like to achieve?
- simply run the two queries in parallel compare the results and return when they both completed (deciding which one to return based on the results) ?
- run the two queries in parallel, return the result when the first is done independently on the comparison result and let the second and the comparison continue after the request has terminated ?
- statically define which query is the source of truth, run both of them in parallel, wait for the source of truth to return, immediately return the result and then do the comparison (potentially waiting for the second query) after the request is done? 
- run the source of truth, return the result and then on a separate thread (detached from the main request) run the second query and compare the result ?

Reply:
the fourth one

Reply:
So I have a little concern with the 4th (which though would be easier to implement). On a long query you may be running the second query a full 15-20 seconds after the first started so, the results won't will start drifting from the results of the first query. Are you ok with this ?

Reply:
sorry, i misspoke earlier, what's implement here is actually more like the second not the fourth. it's only the comparison that continues on another thread after the request is completed, the queries are kicked off simultaneously